### PR TITLE
install path of gem executables; bad gem home warning

### PR DESCRIPTION
**Pull request by pitr-ch**
connect #59

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -84,10 +84,13 @@ lib/ruby/gems/2.3.0/specifications/*
...
 
-# Lib bin
-lib/bin/**
-!lib/bin/gem
-!lib/bin/irb
-!lib/bin/rake
-!lib/bin/rdoc
-!lib/bin/ri
+# Bin directory
+bin/**
```
Comment (eregon) [inline .gitignore]: Just `*` should be enough

File: truffleruby/src/main/ruby/core/rbconfig.rb [hunk 2]
```diff
@@ -85,7 +85,7 @@ module RbConfig
 
   if ruby_home
     libdir = "#{ruby_home}/lib"
-    bindir = "#{libdir}/bin"
+    bindir = "#{ruby_home}/bin"
```
Comment (bjfish) [inline truffleruby/src/main/ruby/core/rbconfig.rb]: The current bindir is causing me issues when testing gems, so I hope this line change goes in soon.

File: README.md [hunk 3]
```diff
@@ -84,9 +84,11 @@ which can be used to get it working.
 The best way to get started with TruffleRuby is via the GraalVM, which includes
 compatible versions of everything you need as well as TruffleRuby.
 
-http://www.oracle.com/technetwork/oracle-labs/program-languages/
+<http://www.oracle.com/technetwork/oracle-labs/program-languages/>
```
Comment (eregon) [inline README.md]: Why the `<` `>`? It seems to make link without as well.

File: README.md [hunk 4]
```diff
@@ -84,9 +84,11 @@ which can be used to get it working.
...
 
 Inside the GraalVM is a `bin/ruby` command that runs TruffleRuby.
+See [`doc/user/using-graalvm`](https://github.com/graalvm/truffleruby/tree/master/doc/user/using-graalvm.md)
```
Comment (eregon) [inline README.md]: `See [Using GraalVM](...`

File: doc/user/installing-gems.md [hunk 5]
```diff
@@ -1,37 +1,30 @@
...
-rvm use system
-rbenv system
-chruby system
-```
-Otherwise, unset the variables with:
-```bash
-unset GEM_HOME GEM_PATH GEM_ROOT
-```
+> **Note:** Ensure you have ruby managers configured properly, see 
+[`doc/user/ruby-managers`](https://github.com/graalvm/truffleruby/blob/master/doc/user/ruby-managers.md).
```
Comment (eregon) [inline doc/user/installing-gems.md]: This should not be a note, it's a critical step.

File: doc/user/installing-gems.md [hunk 6]
```diff
@@ -1,37 +1,30 @@
...
 
 Now install Bundler:
 
-> Bundler `1.14.x` is not yet supported by bundler-workarounds, please use 1.13 
-> in the meanwhile.
+> **Note:** Bundler `1.14.x` is not yet supported by bundler-workarounds, 
+> please use 1.13 in the meanwhile.
 
-    ruby -rbundler-workarounds -S gem install bundler -v 1.13.7
+    ruby -r bundler-workarounds -S gem install bundler -v 1.13.7
```
Comment (eregon) [inline doc/user/installing-gems.md]: I like the one argument `-rlibrary`, it's more convenient to navigate between words at the command line and I believe is the most comment usage over `-r library`.

File: doc/user/installing-gems.md [hunk 7]
```diff
@@ -1,37 +1,30 @@
...
-    ruby -rbundler-workarounds -S bundle install
-    ruby -rbundler-workarounds -S bundle update
+    ruby -r bundler-workarounds -S bundle install
+    ruby -r bundler-workarounds -S bundle update
     
 `bundle exec` does not need the `bundler-workarounds` loaded. 
     
     ruby -S bundle exec bin/rails server
+
+Next step: [Playing Optcarrot](https://github.com/graalvm/truffleruby/blob/master/doc/user/optcarrot.md),
```
Comment (eregon) [inline doc/user/installing-gems.md]: How is this related to `Installing Gems`?

File: doc/user/optcarrot.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# Playing optcarrot
+
+Clone optcarrot repository.
```
Comment (eregon) [inline doc/user/optcarrot.md]: Clone +the+

File: doc/user/optcarrot.md [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+    git clone https://github.com/mame/optcarrot.git
+    cd optcarrot
+
+You will need mplayer to be installed, then you can play Lan Master
```
Comment (eregon) [inline doc/user/optcarrot.md]: `mplayer`

File: doc/user/optcarrot.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+Clone optcarrot repository.
+
+    git clone https://github.com/mame/optcarrot.git
+    cd optcarrot
+
+You will need mplayer to be installed, then you can play Lan Master 
+with following command.
+
+    path/to/graalvm-0.20/bin/ruby bin/optcarrot --video mplayer --input term examples/Lan_Master.nes
```
Comment (eregon) [inline doc/user/optcarrot.md]: Did you test it? I never tried with the upstream `mame/optcarrot`.
There might be issues for input.

File: doc/user/ruby-managers.md [hunk 11]
```diff
@@ -0,0 +1,28 @@
+# Ensuring Ruby managers are configured correctly
+
+If you are using a Ruby manager like `rvm`, `rbnev`, or `chruby` make 
+sure that the manager does not set environment variables `GEM_HOME`, `GEM_PATH`,
+and `GEM_ROOT`. The variables are picked up by truffleruby (as any other Ruby 
+implementation would do) causing truffleruby to pickup wrong gem-home directory 
+instead of it's own.
```
Comment (eregon) [inline doc/user/ruby-managers.md]: +the+ wrong
its own

File: doc/user/ruby-managers.md [hunk 12]
```diff
@@ -0,0 +1,28 @@
+# Ensuring Ruby managers are configured correctly
+
+If you are using a Ruby manager like `rvm`, `rbnev`, or `chruby` make 
+sure that the manager does not set environment variables `GEM_HOME`, `GEM_PATH`,
+and `GEM_ROOT`. The variables are picked up by truffleruby (as any other Ruby
```
Comment (eregon) [inline doc/user/ruby-managers.md]: By RubyGems actually but it might easier to simplify as you did.

File: doc/user/ruby-managers.md [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+Otherwise, unset the variables with:
+
+```bash
+unset GEM_HOME GEM_PATH GEM_ROOT
+```
+
+It will be easier when TruffleRuby can be added to ruby managers, we are working 
+on it.  
+
+Next step: [Installing gems](https://github.com/graalvm/truffleruby/blob/master/doc/user/isntalling-gems.md).
```
Comment (eregon) [inline doc/user/ruby-managers.md]: All links can be relative, GitHub handles it nicely and it means it works on branches (test it with the markdown preview in the PR).
BTW, typo in the link here.

File: truffleruby/src/main/ruby/post-boot/post-boot.rb [hunk 14]
```diff
@@ -16,6 +16,27 @@
     require 'rubygems'
   rescue LoadError
   else
+    unless Gem.dir.include?(Truffle::Boot.ruby_home)
```
Comment (eregon) [inline truffleruby/src/main/ruby/post-boot/post-boot.rb]: I think we should we check the whole `Gem.path`.
Otherwise RubyGems might still load gems from another Ruby.

File: truffleruby/src/main/ruby/post-boot/post-boot.rb [hunk 15]
```diff
@@ -16,6 +16,27 @@
...
+      # rbenv does not set GEM_HOME
+      # rbenv-gemset has to be installed which does set GEM_HOME, it's in the subdir of Truffle::Boot.ruby_home
+      # rbenv/versions/<ruby>/gemsets
+      bad_gem_home ||= Gem.dir.include?('rbenv/versions') && !Gem.dir.include?('rbenv/versions/truffleruby')
+
+      # rvm stores gems at .rvm/gems/<ruby>@<gemset-name>
+      bad_gem_home ||= Gem.dir.include?('rvm/gems') && !Gem.dir.include?('rvm/gems/truffleruby')
+
+      # chruby stores gem in ~/.gem/<ruby>/<version>
+      bad_gem_home ||= Gem.dir.include?('.gem') && !Gem.dir.include?('.gems/truffleruby')
```
Comment (eregon) [inline truffleruby/src/main/ruby/post-boot/post-boot.rb]: Seeing the logic here, how about:
```ruby
our_gem_home = Gem.default_dir
bad_gem_home = dir != our_gem_home && !dir.include?('/truffleruby/')

File: README.md (hunk 3 above)
Comment (pitr-ch) [inline README.md]: Ha, I've confused the syntax with some other markup.

File: doc/user/installing-gems.md (hunk 5 above)
Comment (pitr-ch) [inline doc/user/installing-gems.md]: if the user is following the steps, he has visited the page already and configured it, therefore it was kept as note only. I don't mind either way, changing it.

File: doc/user/installing-gems.md (hunk 6 above)
Comment (pitr-ch) [inline doc/user/installing-gems.md]: It's only simple to write, for reading I find it terrible, the word is joined with the option and there is no space for eyes to find the start of the value. I'll keep spaces.

File: doc/user/installing-gems.md (hunk 7 above)
Comment (pitr-ch) [inline doc/user/installing-gems.md]: The next step is not meant to imply relation but rather what to read next.

File: doc/user/optcarrot.md (hunk 10 above)
Comment (pitr-ch) [inline doc/user/optcarrot.md]: ah, tested but I forgot I am on a branch, I'll have a look

File: doc/user/ruby-managers.md (hunk 12 above)
Comment (pitr-ch) [inline doc/user/ruby-managers.md]: yeah, every ruby has rubygems by default ...

File: truffleruby/src/main/ruby/post-boot/post-boot.rb (hunk 15 above)
Comment (pitr-ch) [inline truffleruby/src/main/ruby/post-boot/post-boot.rb]: Could be, but I wanted to be more precise. I would like the warning not to live for long anyway so we can revisit if it is not removed soon-ish.

File: doc/user/installing-gems.md (hunk 6 above)
Comment (eregon) [inline doc/user/installing-gems.md]: But it's not how users write it, so that's confusing.

File: doc/user/installing-gems.md (hunk 5 above)
Comment (eregon) [inline doc/user/installing-gems.md]: Users should not need to navigate too much to do one thing. It's also always best to be able to point to a single page and have clear prerequisites.

File: doc/user/installing-gems.md (hunk 7 above)
Comment (eregon) [inline doc/user/installing-gems.md]: These docs should be each about one thing, not reading pages ;)

File: doc/user/compatibility.md [hunk 16]
```diff
@@ -112,7 +112,7 @@ provide the same interface as JRuby does for this functionality.
 
 Calling Ruby code from Java is supported by the `PolyglotEngine` API of Truffle.
 
-http://lafo.ssw.uni-linz.ac.at/javadoc/truffle/latest/com/oracle/truffle/api/vm/PolyglotEngine.html
+<http://lafo.ssw.uni-linz.ac.at/javadoc/truffle/latest/com/oracle/truffle/api/vm/PolyglotEngine.html>
```
Comment (eregon) [inline doc/user/compatibility.md]: Extra `<`

File: doc/user/installing-gems.md [hunk 17]
```diff
@@ -1,37 +1,29 @@
 # Installing gems
 
-As mentioned in the 
-[`README`](https://github.com/graalvm/truffleruby/tree/truffle-head/README.md) 
+As mentioned in the [README](README.md)
```
Comment (eregon) [inline doc/user/installing-gems.md]: broken link, should be `../../README.md`

File: truffleruby/src/main/ruby/post-boot/post-boot.rb (hunk 15 above)
Comment (eregon) [inline truffleruby/src/main/ruby/post-boot/post-boot.rb]: The check is not precise though:
> I think we should we check the whole `Gem.path`.
> Otherwise RubyGems might still load gems from another Ruby.

Comment (eregon) [inline truffleruby/src/main/ruby/post-boot/post-boot.rb]: i.e. I think my check is more precise and simpler, so please use it.

File: doc/user/installing-gems.md (hunk 6 above)
Comment (pitr-ch) [inline doc/user/installing-gems.md]: That's an assumption though, I for example write space even on command line. Since this is a documentation, the more readable option with space separation is a better option then a shortcut.

File: doc/user/installing-gems.md (hunk 17 above)
Comment (pitr-ch) [inline doc/user/installing-gems.md]: ah thanks

File: truffleruby/src/main/ruby/post-boot/post-boot.rb (hunk 15 above)
Comment (pitr-ch) [inline truffleruby/src/main/ruby/post-boot/post-boot.rb]: The condition you are suggesting will catch any GEM_HOME change, the old one catches only different GEM_HOME pointing to a gem-home of another ruby in a ruby manager. That is the missing preciseness in the new condition I am referring too. Gem path can be anything, it cannot be easily checked that it's unintentional change and that the path belongs to other ruby. Anyway we need to only detect the common case of misconfigured a ruby manager. We ignore other changes. This will be deleted when truffleruby can be added to ruby managers.

Comment (eregon) [inline truffleruby/src/main/ruby/post-boot/post-boot.rb]: Checking Gem.path will show you the warning is always there,
as `$HOME/.gem/ruby/2.3.0` is always added by RubyGems, unless GEM_PATH is set
And that's an actual problem as it looks through system Ruby gems if there is a system ruby of version 2.3.x.
(For instance, bad extensions, conflicting gem versions, etc)
But probably it is less problematic or likely than a bad value for GEM_HOME, so let's merge for now.
The right fix for this default system Ruby path is changing RUBY_ENGINE.

Comment (pitr-ch) [inline truffleruby/src/main/ruby/post-boot/post-boot.rb]: Yeah agreed, I am already working on the RUBY_ENGINE fix. Looks promising it should not take too long.
